### DISCUSSION TITLE
CFclient is not connecting to xbox one controller

### DISCUSSION THREAD

**Discussion by Ateet-dev** [bitcraze/discussions#1684] [Q&A]
Hi Bitcrazy ppl :)

I have gotten a new crazyflie 2.1+, and have connected successfully with it using CFclient via Crazyradio PA dongle. I am now trying to connect an Xbox One / Xbox series X|S controller and the CFclient is not detecting it. The OS is able recognize the controller, and I can play video games with it on the computer.

The configure input device drop down is not showing the controller in its options. Please help me figure out this issue of connecting the gamepad. Any input example configuration files are appreciated.

I have gone through this page and couldn't figure out how to make it work. https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/development/dev_info_client/#user-configuration-file.

OS is Windows 11.
CF client is running on WSL 2. The usb devices have been attached to the WSL via usbipd commands.

Configure input device dialog:
![image](https://github.com/user-attachments/assets/66bdcbb2-0724-4ff8-9e3f-cbf1fbe99f79)

lsusb output:
![image](https://github.com/user-attachments/assets/05079abe-7cba-4a6d-a05b-e2020f76fb49)

Output of running CFclient
```
INFO:cfclient.gui:Disabling STL printouts
INFO:cfclient.utils.input.inputreaders:Input readers: ['linuxjsdev', 'pysdl2']
INFO:cfclient.utils.input.inputreaders:Successfully initialized [linuxjsdev]
INFO:cfclient.utils.input.inputreaders:Could not initialize [pysdl2]: No SDL2 support on Linux
INFO:cfclient.utils.input.inputinterfaces:Found interfaces: ['leapmotion', 'wiimote', 'zmqpull']
INFO:cfclient.utils.input.inputinterfaces:Could not initialize [leapmotion]: Leap Motion library probably not installed (No module named 'leapsdk')
INFO:cfclient.utils.input.inputinterfaces:Could not initialize [wiimote]: Missing cwiid (wiimote) driver No module named 'cwiid'
INFO:cfclient.utils.config:Dist config read from /home/ateet/crazyflie/drone_env/lib/python3.12/site-packages/cfclient/configs/config.json
INFO:cfclient.utils.config:Config file read from [/home/ateet/.config/cfclient/config.json]
INFO:cfclient.utils.input.inputinterfaces.zmqpull:Biding ZMQ at tcp://127.0.0.1:1212
INFO:cfclient.utils.input.inputinterfaces.zmqpull:Initialized ZMQ
INFO:cfclient.utils.input.inputinterfaces:Successfully initialized [zmqpull]
INFO:OpenGL.acceleratesupport:No OpenGL_accelerate module loaded: No module named 'OpenGL_accelerate'
INFO:cfclient.gui:If startup fails because of "xcb", install dependency with `sudo apt install libxcb-xinerama0`.
Failed to create wl_display (No such file or directory)
qt.qpa.plugin: Could not load the Qt platform plugin "wayland" in "" even though it was found.
INFO:cfclient.utils.zmq_param:Biding ZMQ for parameters at tcp://*:1213
INFO:cfclient.utils.zmq_led_driver:Biding ZMQ for LED driverat tcp://*:1214
INFO:cfclient.utils.input:Using device blacklist [(VirtualBox|VMware)]
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
WARNING: could not determine DPI
WARNING:vispy:could not determine DPI
MESA: error: ZINK: failed to choose pdev
glx: failed to create drisw screen
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cflib.crtp.radiodriver:v0.53 dongle with serial N/A found
WARNING: could not determine DPI
WARNING:vispy:could not determine DPI
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for toolbox_areas
INFO:cfclient.ui.tab_toolbox:No config found for open_toolboxes
INFO:cfclient.ui.tab_toolbox:No config found for open_toolboxes
INFO:cfclient.utils.input:Selected MUX: Normal
INFO:cfclient.ui.main:Role of ZMQ@tcp://127.0.0.1:1212 is Device
INFO:cfclient.utils.input.mux:Adding device ZMQ@tcp://127.0.0.1:1212 to MUX Normal
INFO:cflib.drivers.cfusb:Looking for devices....
INFO:cflib.crtp.prrtdriver:Initialized PRRT driver.
INFO:cflib.crazyflie:Callback->Connection initialized[radio://0/80/2M]
INFO:cflib.crazyflie:We are connected[radio://0/80/2M], request connection setup
INFO:cflib.crazyflie.platformservice:Request _request_protocol_version()
INFO:cflib.crazyflie:Callback->Connected to [radio://0/80/2M]
WARNING:cflib.crazyflie.log:Error no LogEntry to handle id=3
INFO:cflib.crazyflie.platformservice:_crt_service_callback
INFO:cflib.crazyflie.platformservice:Request protocol version
INFO:cflib.crazyflie.platformservice:_platform_callback
INFO:cflib.crazyflie.platformservice:Protocol version (platform): 7
INFO:cflib.crazyflie.toc:TOC for port [5] found in cache
INFO:cflib.crazyflie:Log TOC finished updating
INFO:cflib.crazyflie.mem:4 memories found
INFO:cflib.crazyflie:Memories finished updating
INFO:cflib.crazyflie.toc:TOC for port [2] found in cache
INFO:cflib.crazyflie:Param TOC finished updating
INFO:cflib.crazyflie:Callback->Connection setup finished [radio://0/80/2M]
INFO:cfclient.utils.logconfigreader:Parsing [stabilizer.json]
INFO:cflib.crazyflie.log:Have successfully started logging for id=1
INFO:cflib.crazyflie.log:Have successfully started logging for id=3
INFO:cflib.crazyflie.log:Have successfully started logging for id=4
INFO:cflib.crazyflie:All parameters updated
INFO:cflib.crazyflie:Callback->Connection completed [radio://0/80/2M]
INFO:cfclient.utils.config:Config file saved to [/home/ateet/.config/cfclient/config.json]
INFO:cflib.crazyflie:Closing link
INFO:cflib.crazyflie:Callback->Disconnected from [radio://0/80/2M]
```

Contents of .config/cfclient/config.json
```
{
  "input_device": "ZMQ@tcp://127.0.0.1:1212",
  "link_uri": "radio://0/80/2M",
  "flightmode": "Normal",
  "open_tabs": "Flight Control",
  "trim_pitch": 0.0,
  "slew_limit": 45,
  "slew_rate": 30,
  "trim_roll": 0.0,
  "max_thrust": 80,
  "min_thrust": 25,
  "max_yaw": 200,
  "max_rp": 30,
  "auto_reconnect": false,
  "device_config_mapping": {},
  "enable_debug_driver": false,
  "input_device_blacklist": "(VirtualBox|VMware)",
  "ui_update_period": 100,
  "enable_zmq_input": true,
  "enable_zmq_param": true,
  "enable_zmq_led": true,
  "window_size": [
    1354,
    592
  ],
  "theme": "Default",
  "assistedControl": 0
}
```

I am running Python 3.12.3. Installed CFclient via pip install.

Please help me figure out this issue. Thanks!

**Comment by jeremweb** [comment]
Hi Ateet-dev,

You can try this:

  -  Edit config.json. Replace the first line:
    "input_device": "ZMQ@tcp://127.0.0.1:1212"
    with:
    "input_device": "Xbox Series X - Electric Carbon".

   - Add ZMQ@tcp://127.0.0.1:1212 to the input_device_blacklist, like this:
    "input_device_blacklist": "(VirtualBox|VMware|ZMQ@tcp://127.0.0.1:1212)".

I'm not entirely sure this will work, but it's worth a try :)

**Reply by Ateet-dev** [reply]
Hey jeremweb,
I tried both and it did not help :(, thank you for replying!

**Comment by ataffanel** [comment]
It is unlikely your problem is related to the client: XBox controllers works with the client as long as they are visible and supported by the OS. My guess is that your problem might be related to WSL. XBox controllers require the ```xpad``` driver on Linux and WSL runs a microsoft-compiled version of the linux kernel. I do not know if ```xpad``` is available in wsl.

One way to check is to look at ```sudo dmesg```. When connecting the controller to Linux you should see something like this:
```
[153840.402536] usb 3-2: new full-speed USB device number 3 using xhci_hcd
[153840.557528] usb 3-2: New USB device found, idVendor=045e, idProduct=0b12, bcdDevice= 5.11
[153840.557542] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[153840.557549] usb 3-2: Product: Controller
[153840.557555] usb 3-2: Manufacturer: Microsoft
[153840.557560] usb 3-2: SerialNumber: __redacted__
[153841.167358] input: Microsoft Xbox Series S|X Controller as /devices/pci0000:00/0000:00:08.1/0000:33:00.4/usb3/3-2/3-2:1.0/input/input77
[153841.168366] usbcore: registered new interface driver xpad
```

The two last lines about input device and xpad are the most important, they tell us that the input device is now registered and available to all applications. On my native Linux installation, this is visible by the client:
![image](https://github.com/user-attachments/assets/5ca0b3fa-9039-4cac-9616-4a885a4d3e14)

If ```xpad``` is not available to wsl, you can try to use another gamepad, maybe more regular USB gamepad would work (on 8bitdo gamepads this is called D-Input mode ...). Otherwise you should likely run the client on Windows. It requires the vscode build tools to be installed but when that is done the client can just be pip-installed on Windows without trouble.

**Reply by Ateet-dev** [reply]
Hey ataffanel,
Thank you for replying.
xpad is installed but the dmesg output does not show the last two lines. I'm seeing the rest of the lines though....

The whole setup works in regular windows, but I would probably need all these to work for building and flashing firmware, which seems to be only available through wsl, is that right?

Edit:
tldr: xpad is installed, and after reconnecting controller, dmesg shows messages only until 'SerialNumber'.

**Reply by ToveRumar** [reply]
Hi @Ateet-dev!

If you try running the cfclient directly in windows, does that work?
